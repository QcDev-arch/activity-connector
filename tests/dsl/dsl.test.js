const DslParser = require('../../app/utils/dsl-parser');

/*-------DSL TESTS--------*/

describe('Test for dsl parser error', () => {
  test('Error time with exam', () => {
    expect(() => {
      DslParser.parse(
        'E1 S2-15m',
      );
    }).toThrow();
  });

  test('Error invalid character in front', () => {
    expect(() => {
      DslParser.parse(
        '_E1 S2',
      );
    }).toThrow();
  });

  test('Error no S or F with a quiz', () => {
    expect(() => {
      DslParser.parse(
        'Q1 S1',
      );
    }).toThrow();
  });

  test('Error no invalid character on other lines', () => {
    expect(() => {
      DslParser.parse(
        'Q1 L1F\nq1 S2',
      );
    }).toThrow();
  });

  test('Error no S or F with a homework', () => {
    expect(() => {
      DslParser.parse(
        'H1 L1',
      );
      }).toThrow();
  });

  test('Error invalid time modifier', () => {
    expect(() => {
      DslParser.parse(
        'H1 L1S-5m@5:5',
      );
    }).toThrow();
  });

  test('Error invalid time modifier with comment', () => {
    expect(() => {
      DslParser.parse(
        '#test\nH1 S1F-1h@05',
      );
    }).toThrow();
  });

  test('Error integer', () => {
    expect(() => {
      DslParser.parse(
        'Hf',
      );
    }).toThrow();
  });

  test('Error invalid time modifier', () => {
    expect(() => {
      DslParser.parse(
        'H1 S1F-1w@',
      );
    }).toThrow();
  });
});

test('Test parser basic', () => {
  expect(
    DslParser.parse(
      'E1 S2\nQ1 S1F S2S-30m\nH1 L2F L3S-1d@23:55 L3S-1d@23:55\nQ1 S1F+1h P2S-15m',
    ),
  ).toStrictEqual([
    [],
    [
      {
        activity: 'Exam 1',
        open: {
          activity: 'Seminar 2',
        },
      },
      {
        activity: 'Moodle Quiz 1',
        open: {
          activity: 'Seminar 1',
          modifier: 'end',
        },
        close: {
          activity: 'Seminar 2',
          modifier: 'start',
          time: {
            modifier: '-',
            number: 30,
            type: 'm',
          },
        },
      },
      {
        activity: 'Moodle Homework 1',
        open: {
          activity: 'Laboratory 2',
          modifier: 'end',
        },
        due: {
          activity: 'Laboratory 3',
          modifier: 'start',
          time: {
            modifier: '-',
            number: 1,
            type: 'd',
            at: '23:55',
          },
        },
        cutoff: {
          activity: 'Laboratory 3',
          modifier: 'start',
          time: {
            modifier: '-',
            number: 1,
            type: 'd',
            at: '23:55',
          },
        },
      },
      {
        activity: 'Moodle Quiz 1',
        open: {
          activity: 'Seminar 1',
          modifier: 'end',
          time: {
            modifier: '+',
            number: 1,
            type: 'h',
          },
        },
        close: {
          activity: 'Practicum 2',
          modifier: 'start',
          time: {
            modifier: '-',
            number: 15,
            type: 'm',
          },
        },
      },
    ],
    'EOF',
  ]);
});

test('Test parser with a lot of stuff', () => {
  expect(
    DslParser.parse(
      'E1 S2\nQ1 S1F S2S-30m\nH1 L2F L3S-1d@23:55 L3S-1d@23:55\nQ1 S1F+1h P2S-15m\nQ1 S1F S2S-30m\nH1 L2F L3S-1d@23:55 L3S-1d@23:55\nQ1 S1F+1h P2S-15m\nQ1 S1F S2S-30m\nH1 L2F L3S-1d@23:55 L3S-1d@23:55\nQ1 S1F+1h P2S-15m\nQ1 S1F S2S-30m\nH1 L2F L3S-1d@23:55 L3S-1d@23:55\nQ1 S1F+1h P2S-15m\nQ1 S1F S2S-30m\nH1 L2F L3S-1d@23:55 L3S+1d@23:55\nQ1 S1F+1h P2S-15m\nQ1 S1F S2S-30m\nH1 L2F L3S-1d@23:55 L3S-1d@23:55\nQ1 S1F+1h P2S-15w\nQ1 S1F S2S-30m\nH1 L2F L3S-1d@23:55 L3S-1d@23:55\nQ1 S1F+1h P2S-15m\nE2 S13',
    ),
  ).toStrictEqual([
    [],
    [
      {
        activity: 'Exam 1',
        open: {
          activity: 'Seminar 2',
        },
      },
      {
        activity: 'Moodle Quiz 1',
        open: {
          activity: 'Seminar 1',
          modifier: 'end',
        },
        close: {
          activity: 'Seminar 2',
          modifier: 'start',
          time: {
            modifier: '-',
            number: 30,
            type: 'm',
          },
        },
      },
      {
        activity: 'Moodle Homework 1',
        open: {
          activity: 'Laboratory 2',
          modifier: 'end',
        },
        due: {
          activity: 'Laboratory 3',
          modifier: 'start',
          time: {
            modifier: '-',
            number: 1,
            type: 'd',
            at: '23:55',
          },
        },
        cutoff: {
          activity: 'Laboratory 3',
          modifier: 'start',
          time: {
            modifier: '-',
            number: 1,
            type: 'd',
            at: '23:55',
          },
        },
      },
      {
        activity: 'Moodle Quiz 1',
        open: {
          activity: 'Seminar 1',
          modifier: 'end',
          time: {
            modifier: '+',
            number: 1,
            type: 'h',
          },
        },
        close: {
          activity: 'Practicum 2',
          modifier: 'start',
          time: {
            modifier: '-',
            number: 15,
            type: 'm',
          },
        },
      },
      {
        activity: 'Moodle Quiz 1',
        open: {
          activity: 'Seminar 1',
          modifier: 'end',
        },
        close: {
          activity: 'Seminar 2',
          modifier: 'start',
          time: {
            modifier: '-',
            number: 30,
            type: 'm',
          },
        },
      },
      {
        activity: 'Moodle Homework 1',
        open: {
          activity: 'Laboratory 2',
          modifier: 'end',
        },
        due: {
          activity: 'Laboratory 3',
          modifier: 'start',
          time: {
            modifier: '-',
            number: 1,
            type: 'd',
            at: '23:55',
          },
        },
        cutoff: {
          activity: 'Laboratory 3',
          modifier: 'start',
          time: {
            modifier: '-',
            number: 1,
            type: 'd',
            at: '23:55',
          },
        },
      },
      {
        activity: 'Moodle Quiz 1',
        open: {
          activity: 'Seminar 1',
          modifier: 'end',
          time: {
            modifier: '+',
            number: 1,
            type: 'h',
          },
        },
        close: {
          activity: 'Practicum 2',
          modifier: 'start',
          time: {
            modifier: '-',
            number: 15,
            type: 'm',
          },
        },
      },
      {
        activity: 'Moodle Quiz 1',
        open: {
          activity: 'Seminar 1',
          modifier: 'end',
        },
        close: {
          activity: 'Seminar 2',
          modifier: 'start',
          time: {
            modifier: '-',
            number: 30,
            type: 'm',
          },
        },
      },
      {
        activity: 'Moodle Homework 1',
        open: {
          activity: 'Laboratory 2',
          modifier: 'end',
        },
        due: {
          activity: 'Laboratory 3',
          modifier: 'start',
          time: {
            modifier: '-',
            number: 1,
            type: 'd',
            at: '23:55',
          },
        },
        cutoff: {
          activity: 'Laboratory 3',
          modifier: 'start',
          time: {
            modifier: '-',
            number: 1,
            type: 'd',
            at: '23:55',
          },
        },
      },
      {
        activity: 'Moodle Quiz 1',
        open: {
          activity: 'Seminar 1',
          modifier: 'end',
          time: {
            modifier: '+',
            number: 1,
            type: 'h',
          },
        },
        close: {
          activity: 'Practicum 2',
          modifier: 'start',
          time: {
            modifier: '-',
            number: 15,
            type: 'm',
          },
        },
      },
      {
        activity: 'Moodle Quiz 1',
        open: {
          activity: 'Seminar 1',
          modifier: 'end',
        },
        close: {
          activity: 'Seminar 2',
          modifier: 'start',
          time: {
            modifier: '-',
            number: 30,
            type: 'm',
          },
        },
      },
      {
        activity: 'Moodle Homework 1',
        open: {
          activity: 'Laboratory 2',
          modifier: 'end',
        },
        due: {
          activity: 'Laboratory 3',
          modifier: 'start',
          time: {
            modifier: '-',
            number: 1,
            type: 'd',
            at: '23:55',
          },
        },
        cutoff: {
          activity: 'Laboratory 3',
          modifier: 'start',
          time: {
            modifier: '-',
            number: 1,
            type: 'd',
            at: '23:55',
          },
        },
      },
      {
        activity: 'Moodle Quiz 1',
        open: {
          activity: 'Seminar 1',
          modifier: 'end',
          time: {
            modifier: '+',
            number: 1,
            type: 'h',
          },
        },
        close: {
          activity: 'Practicum 2',
          modifier: 'start',
          time: {
            modifier: '-',
            number: 15,
            type: 'm',
          },
        },
      },
      {
        activity: 'Moodle Quiz 1',
        open: {
          activity: 'Seminar 1',
          modifier: 'end',
        },
        close: {
          activity: 'Seminar 2',
          modifier: 'start',
          time: {
            modifier: '-',
            number: 30,
            type: 'm',
          },
        },
      },
      {
        activity: 'Moodle Homework 1',
        open: {
          activity: 'Laboratory 2',
          modifier: 'end',
        },
        due: {
          activity: 'Laboratory 3',
          modifier: 'start',
          time: {
            modifier: '-',
            number: 1,
            type: 'd',
            at: '23:55',
          },
        },
        cutoff: {
          activity: 'Laboratory 3',
          modifier: 'start',
          time: {
            modifier: '+',
            number: 1,
            type: 'd',
            at: '23:55',
          },
        },
      },
      {
        activity: 'Moodle Quiz 1',
        open: {
          activity: 'Seminar 1',
          modifier: 'end',
          time: {
            modifier: '+',
            number: 1,
            type: 'h',
          },
        },
        close: {
          activity: 'Practicum 2',
          modifier: 'start',
          time: {
            modifier: '-',
            number: 15,
            type: 'm',
          },
        },
      },
      {
        activity: 'Moodle Quiz 1',
        open: {
          activity: 'Seminar 1',
          modifier: 'end',
        },
        close: {
          activity: 'Seminar 2',
          modifier: 'start',
          time: {
            modifier: '-',
            number: 30,
            type: 'm',
          },
        },
      },
      {
        activity: 'Moodle Homework 1',
        open: {
          activity: 'Laboratory 2',
          modifier: 'end',
        },
        due: {
          activity: 'Laboratory 3',
          modifier: 'start',
          time: {
            modifier: '-',
            number: 1,
            type: 'd',
            at: '23:55',
          },
        },
        cutoff: {
          activity: 'Laboratory 3',
          modifier: 'start',
          time: {
            modifier: '-',
            number: 1,
            type: 'd',
            at: '23:55',
          },
        },
      },
      {
        activity: 'Moodle Quiz 1',
        open: {
          activity: 'Seminar 1',
          modifier: 'end',
          time: {
            modifier: '+',
            number: 1,
            type: 'h',
          },
        },
        close: {
          activity: 'Practicum 2',
          modifier: 'start',
          time: {
            modifier: '-',
            number: 15,
            type: 'w',
          },
        },
      },
      {
        activity: 'Moodle Quiz 1',
        open: {
          activity: 'Seminar 1',
          modifier: 'end',
        },
        close: {
          activity: 'Seminar 2',
          modifier: 'start',
          time: {
            modifier: '-',
            number: 30,
            type: 'm',
          },
        },
      },
      {
        activity: 'Moodle Homework 1',
        open: {
          activity: 'Laboratory 2',
          modifier: 'end',
        },
        due: {
          activity: 'Laboratory 3',
          modifier: 'start',
          time: {
            modifier: '-',
            number: 1,
            type: 'd',
            at: '23:55',
          },
        },
        cutoff: {
          activity: 'Laboratory 3',
          modifier: 'start',
          time: {
            modifier: '-',
            number: 1,
            type: 'd',
            at: '23:55',
          },
        },
      },
      {
        activity: 'Moodle Quiz 1',
        open: {
          activity: 'Seminar 1',
          modifier: 'end',
          time: {
            modifier: '+',
            number: 1,
            type: 'h',
          },
        },
        close: {
          activity: 'Practicum 2',
          modifier: 'start',
          time: {
            modifier: '-',
            number: 15,
            type: 'm',
          },
        },
      },
      {
        activity: 'Exam 2',
        open: {
          activity: 'Seminar 13',
        },
      },
    ],
    'EOF',
  ]);
});
