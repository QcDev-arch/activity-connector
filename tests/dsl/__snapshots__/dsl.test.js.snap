// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test parser basic 1`] = `
Array [
  Array [],
  Array [
    Object {
      "activity": "Exam 1",
      "open": Object {
        "activity": "Seminar 2",
      },
    },
    Object {
      "activity": "Moodle Quiz 1",
      "close": Object {
        "activity": "Seminar 2",
        "modifier": "start",
        "time": Object {
          "modifier": "-",
          "number": 30,
          "type": "m",
        },
      },
      "open": Object {
        "activity": "Seminar 1",
        "modifier": "end",
      },
    },
    Object {
      "activity": "Moodle Homework 1",
      "cutoff": Object {
        "activity": "Laboratory 3",
        "modifier": "start",
        "time": Object {
          "at": "23:55",
          "modifier": "-",
          "number": 1,
          "type": "d",
        },
      },
      "due": Object {
        "activity": "Laboratory 3",
        "modifier": "start",
        "time": Object {
          "at": "23:55",
          "modifier": "-",
          "number": 1,
          "type": "d",
        },
      },
      "open": Object {
        "activity": "Laboratory 2",
        "modifier": "end",
      },
    },
    Object {
      "activity": "Moodle Quiz 1",
      "close": Object {
        "activity": "Practicum 2",
        "modifier": "start",
        "time": Object {
          "modifier": "-",
          "number": 15,
          "type": "m",
        },
      },
      "open": Object {
        "activity": "Seminar 1",
        "modifier": "end",
        "time": Object {
          "modifier": "+",
          "number": 1,
          "type": "h",
        },
      },
    },
  ],
  "EOF",
]
`;

exports[`Test parser with a lot of stuff 1`] = `
Array [
  Array [],
  Array [
    Object {
      "activity": "Exam 1",
      "open": Object {
        "activity": "Seminar 2",
      },
    },
    Object {
      "activity": "Moodle Quiz 1",
      "close": Object {
        "activity": "Seminar 2",
        "modifier": "start",
        "time": Object {
          "modifier": "-",
          "number": 30,
          "type": "m",
        },
      },
      "open": Object {
        "activity": "Seminar 1",
        "modifier": "end",
      },
    },
    Object {
      "activity": "Moodle Homework 1",
      "cutoff": Object {
        "activity": "Laboratory 3",
        "modifier": "start",
        "time": Object {
          "at": "23:55",
          "modifier": "-",
          "number": 1,
          "type": "d",
        },
      },
      "due": Object {
        "activity": "Laboratory 3",
        "modifier": "start",
        "time": Object {
          "at": "23:55",
          "modifier": "-",
          "number": 1,
          "type": "d",
        },
      },
      "open": Object {
        "activity": "Laboratory 2",
        "modifier": "end",
      },
    },
    Object {
      "activity": "Moodle Quiz 1",
      "close": Object {
        "activity": "Practicum 2",
        "modifier": "start",
        "time": Object {
          "modifier": "-",
          "number": 15,
          "type": "m",
        },
      },
      "open": Object {
        "activity": "Seminar 1",
        "modifier": "end",
        "time": Object {
          "modifier": "+",
          "number": 1,
          "type": "h",
        },
      },
    },
    Object {
      "activity": "Moodle Quiz 1",
      "close": Object {
        "activity": "Seminar 2",
        "modifier": "start",
        "time": Object {
          "modifier": "-",
          "number": 30,
          "type": "m",
        },
      },
      "open": Object {
        "activity": "Seminar 1",
        "modifier": "end",
      },
    },
    Object {
      "activity": "Moodle Homework 1",
      "cutoff": Object {
        "activity": "Laboratory 3",
        "modifier": "start",
        "time": Object {
          "at": "23:55",
          "modifier": "-",
          "number": 1,
          "type": "d",
        },
      },
      "due": Object {
        "activity": "Laboratory 3",
        "modifier": "start",
        "time": Object {
          "at": "23:55",
          "modifier": "-",
          "number": 1,
          "type": "d",
        },
      },
      "open": Object {
        "activity": "Laboratory 2",
        "modifier": "end",
      },
    },
    Object {
      "activity": "Moodle Quiz 1",
      "close": Object {
        "activity": "Practicum 2",
        "modifier": "start",
        "time": Object {
          "modifier": "-",
          "number": 15,
          "type": "m",
        },
      },
      "open": Object {
        "activity": "Seminar 1",
        "modifier": "end",
        "time": Object {
          "modifier": "+",
          "number": 1,
          "type": "h",
        },
      },
    },
    Object {
      "activity": "Moodle Quiz 1",
      "close": Object {
        "activity": "Seminar 2",
        "modifier": "start",
        "time": Object {
          "modifier": "-",
          "number": 30,
          "type": "m",
        },
      },
      "open": Object {
        "activity": "Seminar 1",
        "modifier": "end",
      },
    },
    Object {
      "activity": "Moodle Homework 1",
      "cutoff": Object {
        "activity": "Laboratory 3",
        "modifier": "start",
        "time": Object {
          "at": "23:55",
          "modifier": "-",
          "number": 1,
          "type": "d",
        },
      },
      "due": Object {
        "activity": "Laboratory 3",
        "modifier": "start",
        "time": Object {
          "at": "23:55",
          "modifier": "-",
          "number": 1,
          "type": "d",
        },
      },
      "open": Object {
        "activity": "Laboratory 2",
        "modifier": "end",
      },
    },
    Object {
      "activity": "Moodle Quiz 1",
      "close": Object {
        "activity": "Practicum 2",
        "modifier": "start",
        "time": Object {
          "modifier": "-",
          "number": 15,
          "type": "m",
        },
      },
      "open": Object {
        "activity": "Seminar 1",
        "modifier": "end",
        "time": Object {
          "modifier": "+",
          "number": 1,
          "type": "h",
        },
      },
    },
    Object {
      "activity": "Moodle Quiz 1",
      "close": Object {
        "activity": "Seminar 2",
        "modifier": "start",
        "time": Object {
          "modifier": "-",
          "number": 30,
          "type": "m",
        },
      },
      "open": Object {
        "activity": "Seminar 1",
        "modifier": "end",
      },
    },
    Object {
      "activity": "Moodle Homework 1",
      "cutoff": Object {
        "activity": "Laboratory 3",
        "modifier": "start",
        "time": Object {
          "at": "23:55",
          "modifier": "-",
          "number": 1,
          "type": "d",
        },
      },
      "due": Object {
        "activity": "Laboratory 3",
        "modifier": "start",
        "time": Object {
          "at": "23:55",
          "modifier": "-",
          "number": 1,
          "type": "d",
        },
      },
      "open": Object {
        "activity": "Laboratory 2",
        "modifier": "end",
      },
    },
    Object {
      "activity": "Moodle Quiz 1",
      "close": Object {
        "activity": "Practicum 2",
        "modifier": "start",
        "time": Object {
          "modifier": "-",
          "number": 15,
          "type": "m",
        },
      },
      "open": Object {
        "activity": "Seminar 1",
        "modifier": "end",
        "time": Object {
          "modifier": "+",
          "number": 1,
          "type": "h",
        },
      },
    },
    Object {
      "activity": "Moodle Quiz 1",
      "close": Object {
        "activity": "Seminar 2",
        "modifier": "start",
        "time": Object {
          "modifier": "-",
          "number": 30,
          "type": "m",
        },
      },
      "open": Object {
        "activity": "Seminar 1",
        "modifier": "end",
      },
    },
    Object {
      "activity": "Moodle Homework 1",
      "cutoff": Object {
        "activity": "Laboratory 3",
        "modifier": "start",
        "time": Object {
          "at": "23:55",
          "modifier": "+",
          "number": 1,
          "type": "d",
        },
      },
      "due": Object {
        "activity": "Laboratory 3",
        "modifier": "start",
        "time": Object {
          "at": "23:55",
          "modifier": "-",
          "number": 1,
          "type": "d",
        },
      },
      "open": Object {
        "activity": "Laboratory 2",
        "modifier": "end",
      },
    },
    Object {
      "activity": "Moodle Quiz 1",
      "close": Object {
        "activity": "Practicum 2",
        "modifier": "start",
        "time": Object {
          "modifier": "-",
          "number": 15,
          "type": "m",
        },
      },
      "open": Object {
        "activity": "Seminar 1",
        "modifier": "end",
        "time": Object {
          "modifier": "+",
          "number": 1,
          "type": "h",
        },
      },
    },
    Object {
      "activity": "Moodle Quiz 1",
      "close": Object {
        "activity": "Seminar 2",
        "modifier": "start",
        "time": Object {
          "modifier": "-",
          "number": 30,
          "type": "m",
        },
      },
      "open": Object {
        "activity": "Seminar 1",
        "modifier": "end",
      },
    },
    Object {
      "activity": "Moodle Homework 1",
      "cutoff": Object {
        "activity": "Laboratory 3",
        "modifier": "start",
        "time": Object {
          "at": "23:55",
          "modifier": "-",
          "number": 1,
          "type": "d",
        },
      },
      "due": Object {
        "activity": "Laboratory 3",
        "modifier": "start",
        "time": Object {
          "at": "23:55",
          "modifier": "-",
          "number": 1,
          "type": "d",
        },
      },
      "open": Object {
        "activity": "Laboratory 2",
        "modifier": "end",
      },
    },
    Object {
      "activity": "Moodle Quiz 1",
      "close": Object {
        "activity": "Practicum 2",
        "modifier": "start",
        "time": Object {
          "modifier": "-",
          "number": 15,
          "type": "w",
        },
      },
      "open": Object {
        "activity": "Seminar 1",
        "modifier": "end",
        "time": Object {
          "modifier": "+",
          "number": 1,
          "type": "h",
        },
      },
    },
    Object {
      "activity": "Moodle Quiz 1",
      "close": Object {
        "activity": "Seminar 2",
        "modifier": "start",
        "time": Object {
          "modifier": "-",
          "number": 30,
          "type": "m",
        },
      },
      "open": Object {
        "activity": "Seminar 1",
        "modifier": "end",
      },
    },
    Object {
      "activity": "Moodle Homework 1",
      "cutoff": Object {
        "activity": "Laboratory 3",
        "modifier": "start",
        "time": Object {
          "at": "23:55",
          "modifier": "-",
          "number": 1,
          "type": "d",
        },
      },
      "due": Object {
        "activity": "Laboratory 3",
        "modifier": "start",
        "time": Object {
          "at": "23:55",
          "modifier": "-",
          "number": 1,
          "type": "d",
        },
      },
      "open": Object {
        "activity": "Laboratory 2",
        "modifier": "end",
      },
    },
    Object {
      "activity": "Moodle Quiz 1",
      "close": Object {
        "activity": "Practicum 2",
        "modifier": "start",
        "time": Object {
          "modifier": "-",
          "number": 15,
          "type": "m",
        },
      },
      "open": Object {
        "activity": "Seminar 1",
        "modifier": "end",
        "time": Object {
          "modifier": "+",
          "number": 1,
          "type": "h",
        },
      },
    },
    Object {
      "activity": "Exam 2",
      "open": Object {
        "activity": "Seminar 13",
      },
    },
  ],
  "EOF",
]
`;
